src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp

```diff
@@ -538,27 +546,41 @@ class KrnlEntryPointOpLowering : public OpRewritePattern<KrnlEntryPointOp> {
       }
     }
 
-    // Create wrapped output.
-    auto wrappedOutput = callApi(
-        rewriter, loc, apiRegistry, API::CREATE_ORDERED_DYN_MEM_REF_DICT, {});
+    auto numOutput = rewriter.create<LLVM::ConstantOp>(
+        loc, int32Ty, rewriter.getI64IntegerAttr(outMemRefList.size()));
+    auto outRmrPtrsArr = rewriter.create<LLVM::AllocaOp>(
+        loc, opaquePtrTy.getPointerTo(), numOutput, /*alignment=*/0);
 
     for (decltype(numOutputs) i = 0; i < outMemRefList.size(); i++) {
       // Get the i-th memref returned, convert to a dynamic memref and store it
       // in the wrappedOutput.
+
       auto memRef = outMemRefList.at(i);
       auto outMemRefTy = memRef.getType().dyn_cast<LLVMType>();
       auto outMemRefRank = getRankFromMemRefType(outMemRefTy);
       auto outMemRefRankVal = rewriter.create<LLVM::ConstantOp>(
           loc, int32Ty, rewriter.getI32IntegerAttr(outMemRefRank));
-      auto outRtMemRef = callApi(rewriter, loc, apiRegistry,
-          API::CREATE_DYN_MEM_REF, {outMemRefRankVal});
+      auto outRtMemRef = callApi(
+          rewriter, loc, apiRegistry, API::CREATE_RTMEMREF, {outMemRefRankVal});
       fillRtMemRefWithMemRef(
           memRef, outRtMemRef, rewriter, loc, apiRegistry, module);
-      auto idx = rewriter.create<LLVM::ConstantOp>(
+
+      auto idxVal = rewriter.create<LLVM::ConstantOp>(
           loc, int32Ty, rewriter.getI32IntegerAttr(i));
-      callApi(rewriter, loc, apiRegistry, API::SET_DYN_MEM_REF,
-          {wrappedOutput, idx, outRtMemRef});
+
+      auto rmrPtrAddrTy = opaquePtrTy.getPointerTo();
+      auto rmrPtrAddr = rewriter
+                            .create<LLVM::GEPOp>(loc, rmrPtrAddrTy,
+                                outRmrPtrsArr, ArrayRef<Value>({idxVal}))
+                            .getResult();
+
+      rewriter.create<LLVM::StoreOp>(loc, outRtMemRef, rmrPtrAddr);
     }
+
+    // Create wrapped output.
+    auto wrappedOutput = callApi(rewriter, loc, apiRegistry,
+        API::CREATE_ORDERED_RTMEMREF_DICT, {outRmrPtrsArr, numOutput});
+
```

Comment:
> both memRef and outRtMemRef have _owningData=true

This is not exactly the case, RtMemRefs are user-facing wrappers for MemRef and regular MemRefs are compiler internal constructs that have no concept of data ownership; part of the reason is that memory management can get complicated within the compiler and data ownership may be too simplified a model to determine malloc/dealloc responsibility. The good thing is that the memory management for regular MemRef is done by the compiler, presumably, it is or will be done in ways that are as optimal as it can be according to static analysis, so we don't really need to worry about managing MemRef's memory (other than making sure that it is not released by the time we need to transfer data out of it) and by extension of this reason, we needn't worry about its data ownership.

In other words, the concept of ownership disappears as we enter the inference function and re-emerges as we exit the function. It only needs to exist where we cannot manage memories automatically.

Reply:
OK as long as we know what happens with `memRef` (i.e., compiler should manage it), that's fine.